File: src/main/java/com/pfe/elearning/quiz/service/QuizService.java
package com.pfe.elearning.quiz.service;

import com.pfe.elearning.question.dto.QuestionCorrect;
import com.pfe.elearning.question.dto.QuestionResponse;
import com.pfe.elearning.question.dto.SubmitAnswerRequest;
import com.pfe.elearning.question.dto.SubmitAnswerResponse;
import com.pfe.elearning.quiz.dto.QuizRequest;
import com.pfe.elearning.quiz.dto.QuizResponse;
import jakarta.transaction.Transactional;
import org.springframework.http.ResponseEntity;

import java.util.List;

public interface QuizService {
    QuizResponse createQuiz(QuizRequest request, Integer courseId);

    QuizResponse createQuizWithQuestions(QuizRequest request, Integer courseId);

    List<QuestionResponse> getAllQuestionsByCourseId(Integer courseId);

    QuizResponse getQuizById(Long quizId);
    QuizResponse updateQuiz(Long quizId, QuizRequest quizRequest);
    void deleteQuiz(Long quizId);
    List<QuizResponse> getAllQuizzes();
    List<QuizResponse> getQuizzesByCourseId(Long courseId);

    List<QuestionCorrect> getQuestionsWithCorrectOptions(Long quizId);

    SubmitAnswerResponse evaluateQuiz(Integer quizId, List<SubmitAnswerRequest> answers, String candidateUsername);

}

File: src/app/services/quiz/quiz.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {environment} from "../../../environments/environment";
import {QuizResponse} from "../../models/quiz/QuizResponse";
import {QuizRequest} from "../../models/quiz/QuizRequest";
import {SubmitAnswerRequest} from "../../models/quiz/SubmitAnswerRequest";
import {userScore} from "../../models/quiz/userScore";
import {SubmitAnswerResponse} from "../../models/quiz/SubmitAnswerResponse";
import {QuizResultResponse} from "../../models/quiz/QuizResultResponse";

@Injectable({
  providedIn: 'root'
})
export class QuizService {

  baseUrl = environment.app.baseUrl;

  constructor(private http: HttpClient) {}

  createQuizForCourse(courseId: number, quiz: QuizRequest): Observable<QuizResponse> {
    return this.http.post<QuizResponse>(`${this.baseUrl}/api/quizzes/course/${courseId}`, quiz);
  }
  createQuizWithQuestions(courseId: number, quiz: QuizRequest): Observable<QuizResponse> {
    return this.http.post<QuizResponse>(`${this.baseUrl}/api/quizzes/course/${courseId}/add`, quiz);
  }

  getQuizById(quizId: number): Observable<QuizResponse> {
    return this.http.get<QuizResponse>(`${this.baseUrl}/api/quizzes/${quizId}`);
  }

  updateQuiz(quizId: number, quiz: QuizRequest): Observable<QuizResponse> {
    return this.http.put<QuizResponse>(`${this.baseUrl}/api/quizzes/${quizId}`, quiz);
  }

  deleteQuiz(quizId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/api/quizzes/${quizId}`);
  }

  getAllQuizzes(): Observable<QuizResponse[]> {
    return this.http.get<QuizResponse[]>(`${this.baseUrl}/api/quizzes/getAll`);
  }
  getQuizzesByCourseId(courseId:number): Observable<QuizResponse[]> {
    return this.http.get<QuizResponse[]>(`${this.baseUrl}/api/quizzes/course/${courseId}`);
  }
  evaluateQuiz(quizId: number, answers: SubmitAnswerRequest[]): Observable<SubmitAnswerResponse> {
    return this.http.post<SubmitAnswerResponse>(`${this.baseUrl}/api/quizzes/${quizId}/evaluateQuiz`, answers);
  }
  getResultsByCandidateId(): Observable<QuizResultResponse[]> {
    return this.http.get<QuizResultResponse[]>(`${this.baseUrl}/api/quizzes/results`);
  }

}
